lib/fetch/response.js

```diff
@@ -369,6 +370,50 @@ function makeNetworkError (reason) {
   })
 }
 
+function makeFilteredResponse (response, state) {
+  state = {
+    internalResponse: response,
+    ...state
+  }
+
+  return new Proxy(response, {
+    get (target, p) {
+      return p in state ? state[p] : target[p]
+    },
+    set (target, p, value) {
+      assert(!(p in state))
+      target[p] = value
+      return true
+    }
+  })
+}
+
+function makeFilteredHeadersList (headersList, filter) {
+  return new Proxy(headersList, {
+    get (target, prop) {
+      // Override methods used by Headers class.
+      if (prop === 'get') {
+        return (name) => filter(name) ? target[prop](name) : undefined
+      } else if (prop === 'has') {
+        return (name) => filter(name) ? target[prop](name) : undefined
+      } else if (prop === 'slice') {
+        return (...args) => {
+          assert(args.length === 0)
+          const arr = []
+          for (let index = 0; index < target.length; index += 2) {
+            if (filter(target[index])) {
+              arr.push(target[index], target[index + 1])
+            }
+          }
+          return arr
+        }
+      } else {
+        assert(false)
+      }
+    }
+  })
+}
```

Comment:
We could just freeze it but I suspect in the future we will need to be able to mutate the headers list for e.g. set-cookie.